pkg/config/structure/unmarshal.go

```diff
@@ -283,7 +287,20 @@ func (n *leafNodeImpl) GetFloat() (float64, error) {
 
 // GetString returns the scalar as a string, or an error otherwise
 func (n *leafNodeImpl) GetString() (string, error) {
-	if n.val.Kind() == reflect.String {
+	switch n.val.Kind() {
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏Do we want to have implicit ITOA ?  I have the feeling that this kind of "auto" type conversion could be error-prone

Reply:
I was originally thinking we'd want to avoid implicitly converting values to string, but we want v1 of the new config to be compatible with the current, so that we don't have to update all of the callers in the codebase that use the config. Once we get better control around the typed behavior of the config, I think it would be worth trying to remove this, but it's not an immediate goal.

Reply:
> Once we get better control around the typed behavior of the config, I think it would be worth trying to remove this, but it's not an immediate goal.

yeah I would prefer to be more permissive and tighten it back down later instead of introducing a regression with the new UnmarshalKey for users with existing configs - here is an example of this being permissive so we need to allow string conversions:

```
(venv) ⬢[jhanna@toolbox datadog-agent]$ sudo cat /etc/datadog-agent/datadog.yaml | grep fleet_layers
fleet_layers: [1, 2.3]
```

```
sudo ./bin/agent/agent diagnose show-metadata inventory-agent

...

        "fleet_policies_applied": [
            "1",
            "2.3"
        ],
```

> 💭 thought: ‏Do we want to have implicit ITOA ? I have the feeling that this kind of "auto" type conversion could be error-prone

I don't think anyone is really writing non-base10 values into their config yaml so imo we're fairly safe with strconv.FormatInt( x, 10) which is what [strconv.Itoa does](https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/strconv/itoa.go;l=35) - viper will parse the yaml to an int or whatever it thinks to be most "accurately" typed and if the struct expects that to convert to a string we have to support it

@misteriaud are there any additional tests you can think of that you'd like to see here?